// eslint-config-rs/index.js
module.exports = {
  extends: [
    'eslint-config-airbnb/',
  ].map(require.resolve),
  env: {
    es6: true,
    browser: true
  },
  parser: "babel-eslint",
  rules: {
    "no-console": [0],
    "space-before-function-paren": [2, "never"],
    "max-len": [2, 120, 2],
    // 不对数组或对象末尾逗号做强制要求
    "comma-dangle": [1, "never"],
    // 不对注释前的空格做强制要求
    "spaced-comment": [0, "always"],
    // 允许使用匿名函数,
    "func-names": 0,
    // 控制逗号前后的空格
    "comma-spacing": [2, {"before": false, "after": true}],
    // 双峰驼命名格式
    "camelcase": 2,
    //大括号内是否允许不必要的空格
    "object-curly-spacing": [0, "never"],
    //换行
    "linebreak-style": [0, "never"],
    //必须使用全等
    "eqeqeq": [0, "never"],
    'react/jsx-filename-extension': [0, "never"],
    'react/prefer-stateless-function': [0, "never"],
    'import/no-extraneous-dependencies': [0, "never"],
    'import/extensions': [0, "never"],
    'import/no-unresolved': [0, "never"],
    'react/prop-types': [0, "never"],
    'import/no-webpack-loader-syntax': [0, "never"],
    'react/jsx-tag-spacing': [0, "never"],
    'react/jsx-space-before-closing': [0, "never"],
    'global-require': [0, "never"],
    "react/jsx-indent": ["error", 2],
    "react/jsx-indent-props": ["error", 2],
    "jsx-a11y/no-static-element-interactions": [0, "never"]
  },
  "ecmaFeatures": {
    "experimentalObjectRestSpread": true
  }
};
